src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -225,14 +197,130 @@ void BaseCouplingScheme::finalize()
 {
   PRECICE_TRACE();
   checkCompletenessRequiredActions();
-  PRECICE_CHECK(isInitialized(), "Called finalize() before initialize()!");
+  PRECICE_CHECK(_initializeHasBeenCalled, "Called finalize() before initialize().");
+}
+
+void BaseCouplingScheme::initialize(double startTime, int startTimeWindow)
+{
+  // Initialize uses the template method pattern (https://en.wikipedia.org/wiki/Template_method_pattern).
+  PRECICE_TRACE(startTime, startTimeWindow);
+  PRECICE_ASSERT(math::greaterEquals(startTime, 0.0), startTime);
+  PRECICE_ASSERT(startTimeWindow >= 0, startTimeWindow);
+  PRECICE_CHECK(not _initializeHasBeenCalled, "initialize() can only be called once.");
+  _time          = startTime;
+  _timeWindows   = startTimeWindow;
+
+  if (_couplingMode == Implicit) {
+    initializeImplicit();
+
+    requireAction(constants::actionWriteIterationCheckpoint());
+    initializeTXTWriters();
+  }
+
+  initializeImplementation();
+
+  if (hasToSendInitData()) {
+    requireAction(constants::actionWriteInitialData());
+  }
+
+  _initializeHasBeenCalled = true;
+}
+
+void BaseCouplingScheme::initializeSendingParticipants(DataMap &dataMap)
+{
+  for (DataMap::value_type &pair : dataMap) {
+    if (pair.second->initialize) {
+      _hasToSendInitData = true;
+      break;
+    }
+  }
+}
+
+void BaseCouplingScheme::initializeReceivingParticipants(DataMap &dataMap)
+{
+  {
+    for (DataMap::value_type &pair : dataMap) {
+      if (pair.second->initialize) {
+        _hasToReceiveInitData = true;
+        break;
+      }
+    }
+  }
+}
+
+
+void BaseCouplingScheme::initializeData()
+{
+  // InitializeData uses the template method pattern (https://en.wikipedia.org/wiki/Template_method_pattern).
+  PRECICE_CHECK(_initializeHasBeenCalled, "initializeData() can be called after initialize() only.");
+  PRECICE_CHECK(not _initializeDataHasBeenCalled, "initializeData() can only be called once.");
+  _initializeDataHasBeenCalled = true;
+  PRECICE_TRACE("initializeData()");
+
+  if (not _hasToSendInitData && not _hasToReceiveInitData) {
+    PRECICE_INFO("initializeData is skipped since no data has to be initialized.");
+    return;
+  }
+
+  PRECICE_DEBUG("Initializing Data ...");
+
+  PRECICE_CHECK(not(_hasToSendInitData && isActionRequired(constants::actionWriteInitialData())),
+                "InitialData has to be written to preCICE before calling initializeData().");
+
+  _hasDataBeenExchanged = false;
+
+  initializeDataImpl();
+}
+
+void BaseCouplingScheme::advance()
+{
+  PRECICE_TRACE(getTimeWindows(), getTime());
+  checkCompletenessRequiredActions();
+  PRECICE_CHECK(_initializeHasBeenCalled, "Before calling advance() coupling has to be initialized via initialize(). This will cause an error in future releases.")  // TODO: preCICE v3.0.0 -> PRECICE_CHECK
```

Comment:
I don't understand the message. It's an error already now.

Reply:
My fault. I actually wanted to use `PRECICE_WARN`. I added the message `"This will cause an error in future releases."`, since I did not want to break existing code (with broken control flow) by throwing an error.

However, I might was missing something here: We already thrown an error, if initialize has not been called on a different level than `BaseCouplingScheme`'s `advance`. I checked what happens, if I comment out `initialize` in `solverdummy.cpp` and indeed an error is thrown. This means that in `BaseCouplingScheme`'s `advence` a `PRECICE_ASSERT` would be appropriate (another level of security does not harm). I changed this correspondingly in a99c820fe6c27a6a38bc1cb5846ca0268ea573ea.

@uekerman Do you think this resolves this conversation?